src/checker.rs

```diff
@@ -15,62 +21,162 @@ pub enum FailureReason {
     DnsError(String),
     Error(String),
 }
+
 /// Fetches the response from a URL.
 ///
 /// First attempts to fetch just the head, and if not supported falls back to fetching the entire body.
-async fn do_request(client: &reqwest::Client, url: &str) -> Result<Response, reqwest::Error> {
-    let response = client.head(url).send().await?;
-    match response.status() {
-        StatusCode::METHOD_NOT_ALLOWED => client.get(url).send().await,
-        _ => Ok(response),
+async fn do_request(
+    client: &reqwest::Client,
+    url: &str,
+) -> Result<(RequestType, Response), (RequestType, reqwest::Error)> {
+    let response = client
+        .head(url)
+        .send()
+        .await
+        .map_err(|e| (RequestType::Head, e))?;
+
+    // Successful head request
+    if response.status() != StatusCode::METHOD_NOT_ALLOWED {
+        return Ok((RequestType::Head, response));
+    }
+
+    client
+        .get(url)
+        .send()
+        .await
+        .map(|r| (RequestType::Get, r))
+        .map_err(|e| (RequestType::Get, e))
+}
+
+trait DnsError {
+    /// Checks if the request error was a DNS error and returns the error string if so
+    fn dns_error(&self) -> Option<String>;
+}
+
+impl DnsError for reqwest::Error {
+    fn dns_error(&self) -> Option<String> {
+        let mut inner = &self as &dyn Error;
+        while let Some(source) = inner.source() {
+            inner = source;
+
+            // TODO: Would be better to get specific errors without string matching like this
+            // Not sure if there's a better way
+            let inner_message = inner.to_string();
+            if inner_message.contains("dns error") {
+                return Some(inner.source().unwrap().to_string());
+            }
+        }
+        None
     }
 }
```

Comment:
No reason to introduce a trait for this, this can just be a function `is_dns_error(&reqwest::Error) -> Option<String>.`